stl/inc/limits

```diff
@@ -1013,20 +1015,83 @@ public:
 };
 
 #ifdef __cpp_lib_bitops // TRANSITION, VSO-1020212
+
+inline bool _Is_tzcnt_supported() {
+    // tzcnt and lzcnt support are indicated with
+    // different cpuid flags.
+    int cpuinfo[4];
+    __cpuid(cpuinfo, 0x07);
+    return cpuinfo[1] & (1 << 3);
+}
+
+
+// see "Hacker's Delight" section 5-3
+template <typename _Ty>
+constexpr int _Countl_zero_helper(_Ty _Val) {
+    _Ty _Yy = 0;
+
+    unsigned int _Nn = numeric_limits<_Ty>::digits;
+    unsigned int _Cc = numeric_limits<_Ty>::digits / 2;
+    do {
+        _Yy = static_cast<_Ty>(_Val >> _Cc);
+        if (_Yy != 0) {
+            _Nn  = _Nn - _Cc;
+            _Val = _Yy;
+        }
+        _Cc = _Cc >> 1;
+    } while (_Cc != 0);
+    return static_cast<int>(_Nn) - static_cast<int>(_Val);
+}
+
+// see "Hacker's Delight" section 5-4
+template <typename _Ty>
+constexpr int _Countr_zero_helper(_Ty _Val) {
+    return numeric_limits<_Ty>::digits
+           - _Countl_zero_helper(static_cast<_Ty>(static_cast<_Ty>(~_Val) & static_cast<_Ty>(_Val - 1)));
+}
+#pragma warning(push)
+#pragma warning(disable : 4640)
+#if defined(_M_IX86) || defined(_M_X64)
+template <typename _Ty>
+inline int _Checked_x86_x86_64_countr_zero(_Ty _Val) {
+    static bool _Have_tzcnt = _Is_tzcnt_supported();
+    constexpr int _Digits   = numeric_limits<_Ty>::digits;
+    constexpr _Ty _Max      = (numeric_limits<_Ty>::max)();
+    if (!_Have_tzcnt && _Val == 0) {

```

Comment:
Why is this really needed?

Reply:
1. With not `_Have_tzcnt` , `__tzcnt` will not work correctly for 0, and *only* for 0
2. Checking first for `_Have_tzcnt` then for 0 instead of jsut for 0 is better for branch prediction, if   `_Have_tzcnt` is true.

Reply:
@AlexGuteniev is exactly correct. tzcnt is just another name for bsr if it's not actually supported, and that check is much, much, easier to optimize than _Val==0.

Reply:
Perhaps a comment would be warranted.